buildpack/telemetry/dynatrace.py

```diff
@@ -1,27 +1,181 @@
 """
-For Dynatrace, metrics are directly ingested through telegraf.
-No additional setup is needed.
-This module only collects information for telegraf from environment variables.
+For Dynatrace, we have two different ingestion methods:
+1. via Dynatrace OneAgent. It's being downloaded and injected to the java
+runtime.
+2. via telegraf. Telegraf ingests custom runtime metrics using Dynatrace
+output plugin
 """
 import logging
 import os
+import json
+from functools import lru_cache
 from urllib.parse import urljoin
 
-INGEST_ENDPOINT = "/api/v2/metrics/ingest"
+from buildpack import util
 
+INGEST_ENDPOINT = "api/v2/metrics/ingest"
+NAMESPACE = "dynatrace"  # to be used to form installation_path global variable
 
-def is_enabled():
+# Environment variables for Dynatrace OneAgent
+# Only passed to the agent if set as environment variable
+default_env = {
+    # -- Environment variables for the integration
+    # "DT_PAAS_TOKEN": required, also used for telegraf integration
+    # "DT_SAAS_URL": required, also used for telegraf integration
+    "DT_TENANT": None,  # required for agent integration, dynatrace environment ID
+    "DT_TENANTTOKEN": None,  # optional, default value is get from manifest.json which is downloaded along with the agent installer
+    # -- Environment variables for orchestration
+    "DT_LOCALTOVIRTUALHOSTNAME": None,  # optional, default not set
+    "DT_APPLICATIONID": None,  # optional, default not set
+    "DT_NODE_ID": None,  # optional, default not set
+    "DT_CLUSTER_ID": None,  # optional, default not set
+    "DT_TAGS": None,  # optional tags e.g. MikesStuff easyTravel=Mike
+    "DT_CUSTOM_PROP": None,  # optional metadata e.g. Department=Acceptance Stage=Sprint
+    # -- Environment variables for troubleshooting
+    "DT_LOGSTREAM": "stdout",  # optional
+    "DT_LOGLEVELCON": None,  # Use this environment variable to define the console log level. Valid options are: NONE, SEVERE, and INFO.
+    "DT_AGENTACTIVE": None,  # Set to true or false to enable or disable OneAgent.
+    # -- Networking environment variables
+    "DT_NETWORK_ZONE": None,  # optional, Specifies to use a network zone. For more information
+    "DT_PROXY": None,  # Optional, When using a proxy, use this environment variable to pass proxy credentials.
+}
+
+
+def stage(buildpack_dir, build_path, cache_path):
+    """
+    Downloads and unzips necessary OneAgent components
+    """
+    if is_agent_enabled():
+        global installation_path
+        installation_path = os.path.join(build_path, NAMESPACE)
```

Comment:
Don't you have to define the var at the top once, before overriding the global var 🤔 And It helps in visibility as a top level var.

Reply:
had to abandon global var approach because path during staging was different than the path during runtime